### ISSUE TITLE
Torchscript access from Libtorch.

### ISSUE THREAD

**Issue by jackeown** [pyg-team/pytorch_geometric#1881]
## ❓ Questions & Help

I saw recently that pytorch geometric now supports torchscript (at least a bit)! :tada:
This is exciting to me since this was the main reason why I hadn't used it about a year ago.
My custom graph neural net code is very slow right now and I'm hoping to use pytorch geometric to make a better implementation.
I'm currently able to use torchscript to export my old model for importing in libtorch from C++, but I'm wondering how this could work with things like pytorch geometric `Data` objects...Although I saw elsewhere that pytorch geometric is "python only", **does this mean that torchscripted pytorch geometric models can't be used from libtorch in C++?** What if I committed to have the jitted models only take in standard libtorch types?

Thanks,
Jack

**Comment by rusty1s** [comment]
Currently, `Data` is not supported by TorchScript and we wait for better custom class TorchScript support to be able to convert it. However, besides data handling, you can now convert your PyG models to TorchScript (and therefore import it in libtorch). However, instead of passing the `data` method to the model's `forward` call, you need to pass in the data attributes explicitly, *e.g.*:
```python
def forward(self, edge_index: Tensor, x: Tensor):
     ...
```

**Comment by christinezuzart** [comment]
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch_geometric.nn as pyg_nn

class GNN(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim, train=True):
        super(GNN, self).__init__()
        self.convs = nn.ModuleList()
        self.convs.append(self.build_conv_model(input_dim, hidden_dim))
        self.lns = nn.ModuleList()
        self.lns.append(nn.LayerNorm(hidden_dim))
        self.lns.append(nn.LayerNorm(hidden_dim))

        for l in range(2):
            self.convs.append(self.build_conv_model(hidden_dim, hidden_dim))

        #post message-passing
        self.post_mp = nn.Sequential(nn.Linear(hidden_dim, hidden_dim), nn.Dropout(0.25), nn.Linear(hidden_dim, output_dim))
        self.dropout = 0.25
        self.num_layers = 3
        self.train = train

    def build_conv_model(self, input_dim, hidden_dim):
        return pyg_nn.GINConv(nn.Sequential(nn.Linear(input_dim,hidden_dim), nn.ReLU(), nn.Linear(hidden_dim, hidden_dim)))

  
    def forward(self, x, edge_index, batch):

        x = self.convs[0](x, edge_index)
        emb = x
        x = F.relu(x)
        x = F.dropout(x, p=self.dropout, training=self.train)
        x = self.lns[0](x)

        x = self.convs[1](x, edge_index)
        emb = x
        x = F.relu(x)
        x = F.dropout(x, p=self.dropout, training=self.train)
        x = self.lns[1](x)

        x = self.convs[2](x, edge_index)
        emb = x
        x = F.relu(x)
        x = F.dropout(x, p=self.dropout, training=self.train)

        x = pyg_nn.global_mean_pool(x, batch)
        higher_emb = x
        x = self.post_mp(x)
        
        return F.log_softmax(x, dim=1)


    def loss(self, pred, label):
        return F.nll_loss(pred, label)

---------------------------------------------------------------------------------
# model -> trained model

model = GNN(num_node_features, hidden_dim, num_classes, train=False)
model.eval = True
with torch.jit.optimized_execution(True):
    traced_script_module = torch.jit.trace(model, (x, edge_index, batch))
-------------------------------------------------------------------------------------
I am getting the below error.Will appreciate any inputs here.

TracerWarning: Output nr 1. of the traced function does not match the corresponding output of the Python function. Detailed error:    
With rtol=1e-05 and atol=1e-05, found 2 element(s) (out of 2) whose difference(s) exceeded the margin of error (including 
0 nan comparisons). The greatest difference was 0.5916104316711426 (-6.0773491859436035 vs. -5.485738754272461), which occurred at index (0, 0).
  _module_class,

**Comment by rusty1s** [comment]
Looks to be a problem with `dropout` in conjunction with `jit.trace`. In particular, you are using PyTorch `eval` mode wrong. You need to remove the `self.train` attribute from the model, and pass `training=self.training` to the dropout layers. Then, you need to set your model into eval mode via `model.eval()`.
